File: src/main/java/cc/mirukuneko/realtrainmodrenewed/client/screen/SignalReceiverScreen.java
package cc.mirukuneko.realtrainmodrenewed.client.screen;

import cc.mirukuneko.realtrainmodrenewed.client.ClientNetworkHelper;
import cc.mirukuneko.realtrainmodrenewed.network.BindSignalReceiverPayload;
import cc.mirukuneko.realtrainmodrenewed.blockentity.SignalRemoteBlockEntity;
import cc.mirukuneko.realtrainmodrenewed.blockentity.SignalStateBlockEntity;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.components.EditBox;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;

public class SignalReceiverScreen extends Screen {
    private final BlockPos pos;
    private EditBox channelBox;

    public SignalReceiverScreen(BlockPos pos) {
        super(Component.literal("受信機"));
        this.pos = pos.immutable();
    }

    @Override
    protected void init() {
        int boxWidth = 140;
        int x = (width - boxWidth) / 2;
        int y = height / 2 - 18;
        channelBox = new EditBox(font, x, y, boxWidth, 20, Component.literal("番号"));
        channelBox.setMaxLength(10);
        readCurrentValue();
        addRenderableWidget(channelBox);
        setInitialFocus(channelBox);

        addRenderableWidget(Button.builder(Component.literal("接続"), button -> submit())
            .bounds(width / 2 - 75, y + 30, 70, 20)
            .build());
        addRenderableWidget(Button.builder(Component.translatable("gui.cancel"), button -> onClose())
            .bounds(width / 2 + 5, y + 30, 70, 20)
            .build());
    }

    private void readCurrentValue() {
        Minecraft minecraft = Minecraft.getInstance();
        if (minecraft.level == null) {
            return;
        }
        if (minecraft.level.getBlockEntity(pos) instanceof SignalRemoteBlockEntity blockEntity && blockEntity.getLinkedChannel() > 0) {
            channelBox.setValue(Integer.toString(blockEntity.getLinkedChannel()));
        } else if (minecraft.level.getBlockEntity(pos) instanceof SignalStateBlockEntity blockEntity && blockEntity.getLinkedChannel() > 0) {
            channelBox.setValue(Integer.toString(blockEntity.getLinkedChannel()));
        }
    }

    private void submit() {
        try {
            int channel = Integer.parseInt(channelBox.getValue().trim());
            ClientNetworkHelper.sendToServer(new BindSignalReceiverPayload(pos, channel));
            onClose();
        } catch (NumberFormatException ignored) {
            if (Minecraft.getInstance().player != null) {
                Minecraft.getInstance().player.sendOverlayMessage(Component.literal("番号を入力してください"));
            }
        }
    }

    @Override
    public void extractRenderState(GuiGraphicsExtractor graphics, int mouseX, int mouseY, float partialTick) {
        extractBackground(graphics, mouseX, mouseY, partialTick);
        super.extractRenderState(graphics, mouseX, mouseY, partialTick);
        graphics.centeredText(font, title, width / 2, height / 2 - 40, 0xFFFFFF);
        graphics.centeredText(font, Component.literal("通信機で表示した番号を入力"), width / 2, height / 2 - 28, 0xAAAAAA);
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/client/screen/SignalReceiverScreen.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.client.screen

import cc.mirukuneko.realtrainmodrenewed.client.ClientNetworkHelper
import cc.mirukuneko.realtrainmodrenewed.network.BindSignalReceiverPayload
import cc.mirukuneko.realtrainmodrenewed.blockentity.SignalRemoteBlockEntity
import cc.mirukuneko.realtrainmodrenewed.blockentity.SignalStateBlockEntity
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.components.Button
import net.minecraft.client.gui.components.EditBox
import net.minecraft.client.gui.screens.Screen
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component

open class SignalReceiverScreen(pos: BlockPos) : Screen(Component.translatable("block.realtrainmodrenewed.signal_receiver")) {

    private val pos: BlockPos = pos.immutable()
    private lateinit var channelBox: EditBox

    override fun init() {
        val boxWidth = 140
        val x = (width - boxWidth) / 2
        val y = height / 2 - 18
        channelBox = EditBox(font, x, y, boxWidth, 20, Component.translatable("screen.realtrainmodrenewed.signal.channel"))
        channelBox.setMaxLength(10)
        readCurrentValue()
        addRenderableWidget(channelBox)
        setInitialFocus(channelBox)

        addRenderableWidget(
            Button.builder(Component.translatable("button.realtrainmodrenewed.connect")) { submit() }
                .bounds(width / 2 - 75, y + 30, 70, 20)
                .build()
        )
        addRenderableWidget(
            Button.builder(Component.translatable("gui.cancel")) { onClose() }
                .bounds(width / 2 + 5, y + 30, 70, 20)
                .build()
        )
    }

    private fun readCurrentValue() {
        val minecraft = Minecraft.getInstance()
        val level = minecraft.level ?: return
        val blockEntity = level.getBlockEntity(pos)
        when {
            blockEntity is SignalRemoteBlockEntity && blockEntity.linkedChannel > 0 ->
                channelBox.value = blockEntity.linkedChannel.toString()
            blockEntity is SignalStateBlockEntity && blockEntity.linkedChannel > 0 ->
                channelBox.value = blockEntity.linkedChannel.toString()
        }
    }

    private fun submit() {
        try {
            val channel = channelBox.value.trim().toInt()
            ClientNetworkHelper.sendToServer(BindSignalReceiverPayload(pos, channel))
            onClose()
        } catch (_: NumberFormatException) {
            Minecraft.getInstance().player?.sendOverlayMessage(
                Component.translatable("message.realtrainmodrenewed.signal.enter_channel")
            )
        }
    }

    override fun extractRenderState(graphics: GuiGraphicsExtractor, mouseX: Int, mouseY: Int, partialTick: Float) {
        super.extractRenderState(graphics, mouseX, mouseY, partialTick)
        graphics.centeredText(font, title, width / 2, height / 2 - 40, 0xFFFFFF)
        graphics.centeredText(font, Component.translatable("screen.realtrainmodrenewed.signal_receiver.hint"), width / 2, height / 2 - 28, 0xAAAAAA)
    }

    override fun isPauseScreen(): Boolean = false
}
